api/controller_test.go

```diff
@@ -545,6 +546,98 @@ func TestController_ListBranchesHandler(t *testing.T) {
 	})
 }
 
+func TestController_ListTagsHandler(t *testing.T) {
+	clt, deps := setupClient(t, "")
+
+	// create user
+	creds := createDefaultAdminUser(t, clt)
+	bauth := httptransport.BasicAuth(creds.AccessKeyID, creds.AccessSecretKey)
+
+	// setup test data
+	ctx := context.Background()
+	_, err := deps.cataloger.CreateRepository(ctx, "repo1", "local://foo1", "master")
+	testutil.Must(t, err)
+	testutil.Must(t, deps.cataloger.CreateEntry(ctx, "repo1", "master", catalog.Entry{Path: "obj1"}))
+	commitLog, err := deps.cataloger.Commit(ctx, "repo1", "master", "first commit", "test", nil)
+	testutil.Must(t, err)
+	const createTagLen = 7
+	var createdTags []*models.Ref
+	for i := 0; i < createTagLen; i++ {
+		tagID := swag.String("tag" + strconv.Itoa(i))
+		commitID := swag.String(commitLog.Reference)
+		_, err := clt.Tags.CreateTag(
+			tags.NewCreateTagParamsWithTimeout(timeout).
+				WithRepository("repo1").
+				WithTag(&models.TagCreation{
+					ID:  tagID,
+					Ref: commitID,
+				}),
+			bauth)
+		testutil.Must(t, err)
+		createdTags = append(createdTags, &models.Ref{
+			ID:       tagID,
+			CommitID: commitID,
+		})
+	}
+
+	t.Run("default", func(t *testing.T) {
+		resp, err := clt.Tags.ListTags(
+			tags.NewListTagsParamsWithTimeout(timeout).
+				WithRepository("repo1").
+				WithAmount(swag.Int64(-1)),
+			bauth)
+		testutil.MustDo(t, "list tags", err)
+		payload := resp.GetPayload()
+		tagsLen := len(payload.Results)
+		if tagsLen != createTagLen {
+			t.Fatalf("ListTags len=%d, expected %d", tagsLen, createTagLen)
+		}
+		if diff := deep.Equal(payload.Results, createdTags); diff != nil {
+			t.Fatal("ListTags results diff:", diff)
+		}
+	})
+
+	t.Run("pagination", func(t *testing.T) {
+		const pageSize = 2
+		var results []*models.Ref
+		var after string
+		var calls int
+		for {
+			calls++
+			resp, err := clt.Tags.ListTags(
+				tags.NewListTagsParamsWithTimeout(timeout).
+					WithRepository("repo1").
+					WithAmount(swag.Int64(pageSize)).
+					WithAfter(swag.String(after)),
+				bauth)
+			testutil.Must(t, err)
+			payload := resp.GetPayload()
+			results = append(results, payload.Results...)
+			if !swag.BoolValue(payload.Pagination.HasMore) {
+				break
+			}
+			after = payload.Pagination.NextOffset
+		}
+		expectedCalls := createTagLen/pageSize + 1
```

Comment:
If `createTagLen` was 8, wouldn't you expect just 4 calls instead of 5?
It's ok to make it const IMO.

Reply:
true - not good code, fixing.